Tests/SourceGenerators/SourceGeneratorAssemblyToProcess/ErrorClasses/UnsupportedRequiredTypes.cs

```diff
@@ -51,11 +51,11 @@ public partial class UnsupportedRequiredTypes : IRealmObject
         [Required]
         public decimal DecimalProp { get; set; }
 
-        [Required]
-        public ObjectId ObjectIdProp { get; set; }
+        // [Required]
+        // public ObjectId ObjectIdProp { get; set; }
```

Comment:
Here is where the main issue lays right now and I am not sure how to tackle it. If you have a property whose type is imported from another module i.e. `ObjectId` is imported from `MongoDB.BSON` then removing the backing fields to the following error:
```
  Fody: An unhandled exception occurred:
Exception:
Member 'MongoDB.Bson.ObjectId <Id>k__BackingField' is declared in another module and needs to be imported
Type:
System.ArgumentException
StackTrace:
   at Mono.Cecil.MetadataBuilder.LookupToken(IMetadataTokenProvider provider) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 2264
   at Mono.Cecil.Cil.CodeWriter.WriteOperand(Instruction instruction) in C:\projects\fody\cecil\Mono.Cecil.Cil\CodeWriter.cs:line 270
   at Mono.Cecil.Cil.CodeWriter.WriteInstructions() in C:\projects\fody\cecil\Mono.Cecil.Cil\CodeWriter.cs:line 180
   at Mono.Cecil.Cil.CodeWriter.WriteResolvedMethodBody(MethodDefinition method) in C:\projects\fody\cecil\Mono.Cecil.Cil\CodeWriter.cs:line 118
   at Mono.Cecil.Cil.CodeWriter.WriteMethodBody(MethodDefinition method) in C:\projects\fody\cecil\Mono.Cecil.Cil\CodeWriter.cs:line 54
   at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1675
   at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1668
   at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1468
   at Mono.Cecil.MetadataBuilder.AddTypes() in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1440
   at Mono.Cecil.MetadataBuilder.BuildTypes() in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1289
   at Mono.Cecil.MetadataBuilder.BuildModule() in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1059
   at Mono.Cecil.MetadataBuilder.BuildMetadata() in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 1029
   at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__2_0(MetadataBuilder builder, MetadataReader _) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 147
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read) in C:\projects\fody\cecil\Mono.Cecil\ModuleDefinition.cs:line 971
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 146
   at Mono.Cecil.ModuleWriter.Write(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 119
   at Mono.Cecil.ModuleWriter.WriteModule(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters) in C:\projects\fody\cecil\Mono.Cecil\AssemblyWriter.cs:line 78
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters) in C:\projects\fody\cecil\Mono.Cecil\ModuleDefinition.cs:line 1160
   at InnerWeaver.WriteModule() in C:\projects\fody\FodyIsolated\ModuleWriter.cs:line 18
   at InnerWeaver.Execute() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 115
Source:
Mono.Cecil
TargetSite:
Mono.Cecil.MetadataToken LookupToken(Mono.Cecil.IMetadataTokenProvider)
```
I am assuming there is something in this imported type reference that refers to the backing field for some reason but after a lot of looking I have not been able to find it. I also tried adding the type of the backing field to `type.NestedTypes` in case it does not actually import the type unless you do this, alas without success.

I am going to try some more tricks but for the sake of testing on this branch right now I commented out the imported fields since doing so does lead to a full re-generation of other classes that do have their backing fields removed! (and seem to work ok!).

Reply:
One thing you could try and do is:

```csharp
var backingFieldName = prop.GetBackingField().FullName;
var backingField = type.Fields.Single(f => f.FullName == backingFieldName);
type.Fields.Remove(backingField);
```

Not sure if that'll work, but might be worth a try.

Reply:
thank for the tip, sadly this does not work 😔 I think the issue here is that it does successfully remove the backing field, it's just that doing so causes some kind of reference to that field which is related to the backing field to fail.
I found some questions loosely related to this: [here](https://stackoverflow.com/questions/38003104/member-httpclient-is-declared-in-another-module-and-needs-to-be-imported-in), [here](https://www.reddit.com/r/csharp/comments/evo5a0/helpmonocecil_how_to_delete_a_method_from_a_type/) and [here](https://stackoverflow.com/questions/32666075/why-mono-cecil-argues-for-method-import-while-i-have-done-it-already). Sadly none of them have direct solutions, I am going to try to manually run through all methods similar to the third link's answer to figure out where the backing field is referred.

Reply:
Okay, I think I can see what the problem is. When you have a class definition like:

```cs
partial class Foo : IRealmObject
{
    public ObjectId { get; set; } = ObjectId.GenerateNewId();
}
```

The field is getting assigned in the object constructor, even if that's not obvious from the code. The generated ctor looks kind of like this:

```
	// Id = ObjectId.GenerateNewId();
	IL_0000: ldarg.0
	IL_0001: call valuetype [MongoDB.Bson]MongoDB.Bson.ObjectId [MongoDB.Bson]MongoDB.Bson.ObjectId::GenerateNewId()
	IL_0006: stfld valuetype [MongoDB.Bson]MongoDB.Bson.ObjectId Realms.Tests.ObjectIdPrimaryKeyWithValueObject::'<Id>k__BackingField'
```

I think the issue manifests with ObjectId/Guid because those are the types we most often assign a default value to, but it'll likely happen for other types as long as you have an assignment in the model. Unfortunately, it's probably going to be a bit tricky to clean those up - we can probably bounce some ideas next week about how it can be achieved.